src/normalizer/mod.rs

```diff
@@ -78,17 +72,127 @@ pub trait Normalizer: Sync + Send {
     /// Normalize the provided [`Token`].
     /// Options can be set using the provided [`NormalizerOption`].
     ///
-    /// A Normalizer can return several `Token`s.
-    fn normalize<'o>(
-        &self,
-        token: Token<'o>,
-        options: NormalizerOption,
-    ) -> Box<dyn Iterator<Item = Token<'o>> + 'o>;
+    fn normalize<'o>(&self, token: Token<'o>, options: NormalizerOption) -> Token<'o>;
+
+    /// Return true if the normalizer can process Token of a specific [`Script`] and [`Language`].
+    ///
+    /// Some normalizer are specialized for a `Script` or/and a `Language` and shouldn't be called on every `Token`s.
+    fn should_normalize(&self, token: &Token) -> bool;
+}
+
+fn shrink_cow<'o>(s: &Cow<'o, str>, new_size: usize) -> Cow<'o, str> {
+    match s {
+        Cow::Borrowed(s) => Cow::Borrowed(&s[..new_size]),
+        Cow::Owned(s) => Cow::Owned(s[..new_size].to_string()),
+    }
+}
+
+pub trait CharNormalizer: Sync + Send {
+    fn normalize_char(&self, c: char) -> Option<CharOrStr>;
+
+    fn normalize_cow_str<'o>(&self, s: Cow<'o, str>) -> Cow<'o, str> {
```

Comment:
Why does `normalize_cow_str` take a `Cow` as argument, as opposed to a `&str`? If I understand correctly, we don't make use of the `Owned` variant at all. But could someone write an alternative implementation that makes use of it?

Reply:
> Why does normalize_cow_str take a Cow as argument, as opposed to a &str?  If I understand correctly, we don't make use of the Owned variant at all

for a token, you can have several normalizers called in a row, so for the first normalizer call, you'll always have a borrowed value, but, this first normalizer could modify the borrowed value, for instance, a lowercase, and creates an Owned value. Then, the next normalizer will have an owned value as a parameter.
Each normalizer will replace the `lemma`, and it is important to note that the previous version will be dropped. So let's play some possible scenarios:
1) `normalizer(1)` returns a `Borrowed(a)` version of `text` then `normalizer(2)` returns a `Borrowed(b)` version of `Borrowed(a)` which is a borrow of `text`, `Borrowed(a)` is dropped, everything is ok.
2) `normalizer(1)` returns an `Owned(a)` version of `text` then `normalizer(2)` returns a `Borrowed(b)` version of `Owned(a)` which is a borrow on a modified clone of `text`, `Owned(a)` is dropped, `Borrowed(b)` can't be valid because of it has a reference on `Owned(a)`.

These cases, force us to know if the passed value is borrowed or not to know if we have to clone it.
However, it's possible that my implementation is not the best, I can just say that it solves this issue. 😬 

> But could someone write an alternative implementation that makes use of it?

Yes but I'll never suggest it